File: app/src/main/java/ch/bailu/aat/services/background/BackgroundTask.java
package ch.bailu.aat.services.background;

import android.content.Context;

import ch.bailu.aat.services.ServiceContext;

public abstract class BackgroundTask implements ThreadControl {

    public static final BackgroundTask NULL = new BackgroundTask() {

        @Override
        public long bgOnProcess(ServiceContext sc) {
            return 0;
        }
    };


    private Exception exception;

    private boolean processing = true;


    @Override
    public boolean canContinue() {
        return processing;
    }



    public abstract long bgOnProcess(ServiceContext sc);


    public synchronized void stopProcessing() {
        processing =false;
    }



    public ThreadControl getThreadControl() {
        return this;
    }

    public void onInsert(Context c) {}
    public void onRemove(Context c) {}

    protected void setException(Exception e) {
        exception = e;
    }

    public  Exception getException() {
        return exception;
    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/service/background/BackgroundTask.kt
package ch.bailu.aat_lib.service.background

import ch.bailu.aat_lib.app.AppContext

abstract class BackgroundTask : ThreadControl {

    private var processing = true

    override fun canContinue(): Boolean {
        return processing
    }

    abstract fun bgOnProcess(appContext: AppContext): Long

    @Synchronized
    fun stopProcessing() {
        processing = false
    }

    val threadControl: ThreadControl
        get() = this

    open fun onInsert() {}
    open fun onRemove() {}

    companion object {
        @JvmField
        val NULL: BackgroundTask = object : BackgroundTask() {
            override fun bgOnProcess(appContext: AppContext): Long {
                return 0
            }
        }
        @JvmField
        val STOP: BackgroundTask = object : BackgroundTask() {
            override fun bgOnProcess(appContext: AppContext): Long {
                return 0
            }

            override fun canContinue(): Boolean {
                return false
            }
        }
    }
}
